### ISSUE TITLE
[Question] Is the densification/pruning logic different from 3DGS vanilla?

### ISSUE THREAD

**Issue by seigeweapon** [yzslab/gaussian-splatting-lightning#29]
hi, by run gaussian-splatting-lightning and 3DGS vanilla code on the same dataset, I got gs models with very different number of points. 

The init point cloud has 2.6M points, it is from a colmap dense reconstruction. After running same number of iterations (10000) on both side. The output of gaussian-splatting-lightning has only 0.37M points, while the original 3DGS has 2.0M points.

I see no difference on the densification duration or other related parameters. And the densification-pruning code seems to be the same. So why the difference?

the command on gaussian-splatting-lightning:
`
python main.py fit --data.path ../../colmap/stegosaurus_colmap/dense/ -n stego_dense_10000 --max_steps 10000
`

and the command on 3DGS:
`
python train.py -s ../../colmap/stegosaurus_colmap/dense/ -r 1 --iterations 10000
`

**Comment by yzslab** [comment]
Theoretically this should not be happened with default hyperparameters.
What is the resolution of your images? Did they resized to 1,600 by vanilla 3DGS?
If resizing happened, try to start vanilla 3DGS training with option `-r 1`: `python train.py -r 1 ...`.

**Comment by seigeweapon** [comment]
updated the post.

the resolution is 3996x2992. I used "-r 1" to make sure no downsampling.

**Comment by yzslab** [comment]
> updated the post.
> 
> the resolution is 3996x2992. I used "-r 1" to make sure no downsampling.

Did they load the same point cloud to do initialization? Check whether both the point number of the `input.ply` file located in the training output directories are identical. Because there are some differences in the point cloud loader.

It would be better if you can send me your dataset, because I can not reproduce this problem.
Send to this address if you do not want to publice it: literal_chamois0i@icloud.com.

**Comment by seigeweapon** [comment]
Dataset sent to you.

Yes, the input.ply are identical. 2.6M points

**Comment by seigeweapon** [comment]
btw, I made some changes to load from ply. I put it here just in case to help you reproduce it.

`
diff --git a/internal/dataparsers/colmap_dataparser.py b/internal/dataparsers/colmap_dataparser.py
index 8731a03..19c1b65 100644
--- a/internal/dataparsers/colmap_dataparser.py
+++ b/internal/dataparsers/colmap_dataparser.py
@@ -6,7 +6,7 @@ from typing import Tuple
 
 import torch
 import numpy as np
-
+from typing import NamedTuple
 from plyfile import PlyData, PlyElement
 
 import internal.utils.colmap as colmap_utils
@@ -16,6 +16,20 @@ from internal.configs.dataset import ColmapParams
 from internal.utils.graphics_utils import getNerfppNorm
 
 
+class BasicPointCloud(NamedTuple):
+    points : np.array
+    colors : np.array
+    normals : np.array
+
+def fetchPly(path):
+    plydata = PlyData.read(path)
+    vertices = plydata['vertex']
+    positions = np.vstack([vertices['x'], vertices['y'], vertices['z']]).T
+    colors = np.vstack([vertices['red'], vertices['green'], vertices['blue']]).T / 255.0
+    normals = np.vstack([vertices['nx'], vertices['ny'], vertices['nz']]).T
+    return BasicPointCloud(points=positions, colors=colors, normals=normals)
+
+
 class ColmapDataParser(DataParser):
     def __init__(self, path: str, output_path: str, global_rank: int, params: ColmapParams) -> None:
         super().__init__()
@@ -194,11 +208,18 @@ class ColmapDataParser(DataParser):
         #         # waiting ply
         #         print("#{} waiting for {}".format(os.getpid(), ply_path))
         #         time.sleep(1)
-        print("loading colmap 3D points")
-        xyz, rgb, _ = ColmapDataParser.read_points3D_binary(
-            os.path.join(sparse_model_dir, "points3D.bin"),
-            selected_image_ids=selected_image_ids,
-        )
+
+        ply_path = os.path.join(sparse_model_dir, "points3D.ply")
+        if os.path.exists(ply_path):
+            print("loading colmap 3D points from points3D.ply")
+            pcd = fetchPly(ply_path)
+            xyz, rgb = pcd.points, pcd.colors
+        else:
+            print("loading colmap 3D points from points3D.bin")
+            xyz, rgb, _ = ColmapDataParser.read_points3D_binary(
+                os.path.join(sparse_model_dir, "points3D.bin"),
+                selected_image_ids=selected_image_ids,
+            )
 
         loaded_mask_count = 0
         # initialize lists

`

**Comment by yzslab** [comment]
> Dataset sent to you.
> 
> Yes, the input.ply are identical. 2.6M points

Received.

Vanilla 3DGS will load `points3D.ply` if exists:
https://github.com/graphdeco-inria/gaussian-splatting/blob/472689c0dc70417448fb451bf529ae532d32c095/scene/dataset_readers.py#L157-L170

While this repo only load `points3D.bin`. Your `points3D.ply` is generated from dense reconstruction, not `poins3D.bin` from sparse reconstruction. So this is the reason why you get the different result.

**Comment by seigeweapon** [comment]
I noticed this. so I made some small changes to enable loading point3D.ply with your code.

**Comment by yzslab** [comment]
> I noticed this. so I made some small changes to enable loading point3D.ply with your code.

OK, I will test it later.

**Comment by yzslab** [comment]
> btw, I made some changes to load from ply. I put it here just in case to help you reproduce it. [ply-diff.txt](https://github.com/yzslab/gaussian-splatting-lightning/files/15261716/ply-diff.txt)

Looks the RGB value from ply file are normalized twice. The first time is in your `fetchPly()`, the second at https://github.com/yzslab/gaussian-splatting-lightning/blob/f577373d301874f9a4cfce6ab24ff85882e5b5f5/internal/dataset.py#L342.

I will test whether the result related to this.

**Comment by yzslab** [comment]
![image](https://github.com/yzslab/gaussian-splatting-lightning/assets/564361/9cda1258-ba08-4317-a201-7223b3fa35ad)

It indeed related to dividing by 255 twice.
Remove the `/ 255.0` operation from your `fetchPly()`, then you should be able to get your expected result.

**Comment by seigeweapon** [comment]
You nailed it. Thank you so much.
